aten/src/ATen/quantized/Quantizer.h

```diff
@@ -0,0 +1,226 @@
+#pragma once
+
+#include <c10/core/QScheme.h>
+#include <c10/macros/Macros.h>
+#include <c10/util/Exception.h>
+#include <ATen/core/Tensor.h>
+
+#include <cmath>
+#include <memory>
+
+// TODO: move to c10 namespace after we
+// unified caffe2::Tensor and at::Tensor
+
+namespace at {
+
+struct QTensorImpl;
+
+using QTensor = Tensor;
+using RealTensor = Tensor;
+
+struct Quantizer;
+using QuantizerPtr = c10::intrusive_ptr<Quantizer>;
+//using QuantizerPtr = std::shared_ptr<Quantizer>;
+
+
+/**
+ * Quantizer is the class for storing all the information
+ * that's necessary to perform quantize and dequantize
+ * operation.
+ *
+ * We might have different types of quantization schemes and this is
+ * the base class for all quantizers.
+ *
+ * QTensorImpl will hold a pointer to Quantizer so that we can support
+ * different quantization schemes on Tensor.
+ *
+ * For example, the most common quantization scheme, Affine Quantization, requires
+ * scale and zero_point as parameters, we'll store scale and zero_point inside
+ * the instance and we can use it to quantize a float Tensor or dequantize a
+ * quantized Tensor.
+ *
+ * When you add new types of leaf Quantizer class, please also
+ * make sure to add a corresponding QScheme enum since
+ * they should have one to one mapping.
+ *
+ * Note about intrusive_ptr:
+ * QTensor holds an intrusive_ptr to Quantizer, and multiple Tensor can
+ * share the same Quantizer.
```

Comment:
what are the semantics of sharing the Quantizer? do we plan to keep them immutable? otherwise updates to values when existing tensors are around might be sketchy. if so, it might be better to define fields as const and initialize them only through constructor.

Reply:
Yes, we plan to keep them immutable